### PR TITLE
Add codecov support

### PR DESCRIPTION
**Pull request by leandro-lucarella-sociomantic**

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -15,4 +15,4 @@
 	url = https://github.com/sociomantic-tsunami/turtle.git
 [submodule "beaver"]
 	path = submodules/beaver
-	url = https://github.com/sociomantic-tsunami/beaver.git
+	url = https://github.com/leandro-lucarella-sociomantic/beaver.git
```
Comment (gavin-norman-sociomantic) [inline .gitmodules]: ?

Comment (leandro-lucarella-sociomantic) [reply .gitmodules]: Just for testing the unreleased beaver that adds support for this (that's why it's WIP and blocked and the commit message says I need to change that before going final :D)

**Comment by leandro-lucarella-sociomantic** [comment]
For some reason the codecov message is not posted, but it seems like it was uploaded correctly:

* https://codecov.io/gh/sociomantic-tsunami/dhtnode
* https://codecov.io/gh/sociomantic-tsunami/dhtnode/pull/75

**Comment by leandro-lucarella-sociomantic** [comment]
FYI @mihails-strasuns-sociomantic

**Comment by leandro-lucarella-sociomantic** [comment]
Updated with new beaver v0.4.0-rc.1 (so no more funny stuff).

**Comment by leandro-lucarella-sociomantic** [comment]
Oh, codecov message is now there!

**Comment by mihails-strasuns-sociomantic** [comment]
Cool that it works but do we actually want it in dhtnode? It has pretty much zero unit tests, and turtle ones don't affect coverage.

**Comment by gavin-norman-sociomantic** [comment]
I also wondered how important code cov is in applications.

**Comment by leandro-lucarella-sociomantic** [comment]
> Cool that it works but do we actually want it in dhtnode? It has pretty much zero unit tests, and turtle ones don't affect coverage.

Why not?

> I also wondered how important code cov is in applications.

I think it good to know how much of our apps is actually tested, why wouldn't that be important?

**Comment by gavin-norman-sociomantic** [comment]
Yeah, fair enough. I guess my trust of this code cov plugin (or whatever it is) is still hovering at around 0%, so I regard it as more of a hindrance than an aid.

**Comment by mihails-strasuns-sociomantic** [comment]
It is additional cognitive load - few lines more to maintain in CI script, tiny bit more noise in pull requests. Not much, but if it doesn't provide any value at all (and I stand by the point that is had exactly zero value for nodes), why make things even tiny bit worse?

Note that _some_ apps do have plenty of unittests though.

**Comment by gavin-norman-sociomantic** [comment]
(I'll leave it for @mihails-strasuns-sociomantic to merge, as he knows more about this.)

**Comment by leandro-lucarella-sociomantic** [comment]
> Yeah, fair enough. I guess my trust of this code cov plugin (or whatever it is) is still hovering at around 0%, so I regard it as more of a hindrance than an aid.

This goes beyond apps, so we have to decide if we trust the service and if we so, use it everywhere when possible, and if not remove it from everywhere. We never left the *experimental* status, but now I think we've used it enough to maybe be able to make that decision. My personal impression is, besides the few problems we had, is useful, in particular for PRs, where you can see very easily how are you impacting the coverage (if you are missing tests).

> if it doesn't provide any value at all

Agreed, but as said before, if it doesn't provide any value, we should remove it from everywhere. Since we have been spending some time on it, I thought it added value (and again, IMHO it does).

I just can't see at all why it should be good for ocean but bad for the dhtnode (or library vs. apps if you will).

**Comment by leandro-lucarella-sociomantic** [comment]
> Note that some apps do have plenty of unittests though.

And again, there is no reason for turtle to not produce coverage (we just need to compile the binary with coverage, if we aren't already). If it's not there yet we can add it (and I think we should).

**Comment by gavin-norman-sociomantic** [comment]
>And again, there is no reason for turtle to not produce coverage (we just need to compile the binary with coverage, if we aren't already). If it's not there yet we can add it (and I think we should).

Yes, that would be very cool.

**Comment by mihails-strasuns** [comment]
> I just can't see at all why it should be good for ocean but bad for the dhtnode (or library vs. apps if you will)

It isn't about apps vs libs. In my opinion only reason to add code coverage metric is if you actually intend to pay attention to it and gradually increase it. This is true for ocean but I expect zero effort on this side in dhtnode in any foreseeable future. Which equates to no value.

> And again, there is no reason for turtle to not produce coverage 

That could be interesting, though I don't yet have an opinion if it is a really helpful metric. The very idea about black box tests is that you don't care (and can use actual production build to test), but I can see how having that kind of information may help.

**Comment by leandro-lucarella-sociomantic** [comment]
> It isn't about apps vs libs. In my opinion only reason to add code coverage metric is if you actually intend to pay attention to it and gradually increase it. This is true for ocean but I expect zero effort on this side in dhtnode in any foreseeable future. Which equates to no value.

Why? Wouldn't you aim at adding coverage if you see some code of a PR is completely untested? I really don't see why we would not want to pay attention to coverage.

> That could be interesting, though I don't yet have an opinion if it is a really helpful metric. The very idea about black box tests is that you don't care (and can use actual production build to test), but I can see how having that kind of information may help.

OK, is true that if we compile with coverage we don't get the exact same binary as used in production, but for me for turtle that's OK, most of the environment is faked anyway so it's not a very realistic testing. I think it would be a worth trade-off to add coverage to turtle tests.

**Comment by leandro-lucarella-sociomantic** [comment]
BTW, judging from the current MakD status, turtle tests should already get coverage. I will updated to the latest MakD to see.

**Comment by mihails-strasuns-sociomantic** [comment]
> Wouldn't you aim at adding coverage if you see some code of a PR is completely untested? 

Nope, definitely not for dhtnode (and that is probably the part you don't realize fully :))

**Comment by leandro-lucarella-sociomantic** [comment]
Definitely. Explain :)

**Comment by mihails-strasuns-sociomantic** [comment]
To be more specific, I would expect 98% to 100% of all dhtnode PRs to have zero unittest coverage

**Comment by leandro-lucarella-sociomantic** [comment]
I'm not talking about unittests, but **any** tests, including turtle.

**Comment by mihails-strasuns-sociomantic** [comment]
Those shouldn't be discussed as some uniform thing. Do you at least see now why I think it makes no sense to track _unit_ test coverage in nodes right now?

**Comment by leandro-lucarella-sociomantic** [comment]
In a way, yes, I'm now seeing that the dhtnode has almost zero testing, but that's bad IMHO. I now (sort of) agree that adding coverage reports **right now** might not be helpful until the test framework of the dht in general is improved, but I would be happy to having something annoying the hell out of us until we get some proper testing for the DHT :P

Have `make test` having only 17% coverage is not.... ideal.

**Comment by leandro-lucarella-sociomantic** [comment]
(and for the records, this PR already uses the latest makd and is getting coverage for `integrationtest` too)

**Comment by mihails-strasuns** [comment]
Ok, now let's talk about turtle because that also answers that adding unit tests to dhtnode should be extremely low priority.

Knowing how much of code is covered by turtle is in general interesting metric to know because it shows how much of dhtnode code is really engaged under regular workload. But at the same time I can imagine few cases where you would pursue increasing that metric deliberately.

Truly important metric for dhtnode test coverage is covered amount of requests and argument permutations for those requests. From a black-box PoV it is the only one that matters because it is only thing any external app can observe. If you have all supported scenarios verified at the network API level there is a high degree of confidence about dhtnode correctness no matter what is internal coverage. I can see value in module integration test for dmqnode for disk overflow feature, but in dhtnode there isn't really any isolated complicated logic that can be verified separately like that.

To sum it up - such information would be useful to know, but of no importance to actively pursue.

**Comment by leandro-lucarella-sociomantic** [comment]
Just some final words because I'm getting tired, and we already discussed about this (how even when there are more important levels for testing, that doesn't mean other levels aren't useful to have, like in this case unittests).

But leaving that aside, code coverage gives you a lower boundary of testing. If your tests are covering 100%, we all know it doesn't mean your code will not have any bugs because coverage only ensures you have gone through one code path at least once, not that you actually tested all the possibilities of what those paths do. But it works the other way around. If you there is code paths that have 0 coverage, that guarantees that path is not tested at all. So if you have perfect tests, it means you have dead code (and probably want to remove it), otherwise it is guaranteed that some code path is not tested.

This is what bothers me. I want to at least know the path is tested. Even when the testing might not be complete, it is at least something. Proving you are testing all cases is almost impossible, but proving you tested a code path at least once tells you at least a bit of how much can you trust your tests.

How you come test the code (if it's unit tests, turtle black-box tests, module integrations tests, or whole system integration tests) is not important (in this context, just in this context), but having at least **something** going through your different code paths for me is a worthy goal.

**Comment by mihails-strasuns-sociomantic** [comment]
I feel so strongly about this because it looks like a manifestation of a cargo cult approach - "add tests because tests are good and higher coverage is always better". Neither of which is always true. It is totally possible to have a decreased coverage in a PR which refactors bunch of code to move it to CTFE. It is totally possible to have tests that provide coverage but don't actually detect any errors.

First question to ask should always be "what error conditions I need to guard against most" and second right after "how may I best achieve that". Pursuing test coverage as its own goal answers neither.

**Comment by leandro-lucarella-sociomantic** [comment]
I agree in general with what you say, but merging this PR is effortless. What you suggest have a much higher cost. I know is not perfect but I think is better than nothing (I guess this is the key part where we don't agree).

But fine, my argumentation ends here, I'll leave it up to you.

**Comment by gavin-norman-sociomantic** [comment]
I see no harm in merging it, now the work has been done.
